### ISSUE TITLE
Is it possible to intercept and modify requests/responses?

### ISSUE THREAD

**Issue by pwntester** [neovim/nvim-lsp#45]
Not sure how to extend LSP for server oddities. This is something required in order to support servers such as eclipse jdt for java. In vim-lsc this can be accomplished with hooks (https://github.com/natebosch/vim-lsc/blob/master/doc/lsc.txt#L446-L492). Is there anything similar in neovim implementation that we can use and configure here?

Thanks

**Comment by norcalli** [comment]
@pwntester for that, you'll want to see the original `:h lsp.txt` docs, but the answer is yes. I designed it to be able to customize it to the nth degree.

It would help if you were more specific in what customization you were trying to achieve so that I can give you a more specific answer.

**Comment by pwntester** [comment]
Thanks for the quick answer, I actually use vim-lsc hooks in the following two cases:

- Java eclipse JDT (see my comment here https://github.com/neovim/nvim-lsp/pull/22#issuecomment-557130911)
In short, JDT use custom commands that need to be changed to conform with the spec. In particular I need to change response to `textDocument/codeAction` so that `java.apply.workspaceEdit` is replaced with `workspaceEdit` (see https://github.com/eclipse/eclipse.jdt.ls/issues/376)

- At least vim-lsc does not send `workspaceFolders` and `rootPath` params in the `initialize` request, since they are expected by some servers I use the following hook to add them (Im basically copying rootUri into them):

```
            \ 'message_hooks': {
            \   'initialize': {
            \       'rootPath': {method, params -> params['rootUri'][7:]},
            \       'workspaceFolders': {method, params -> [{'name': 'test', 'uri': params['rootUri']}]},
            \   }
            \ }
```

**Comment by norcalli** [comment]
Generally, you'd want to consult the `:h lsp.txt` docs to see the notes on `start_client(config)`, which is the main entrypoint for making clients and the thing which the skeletons here generate based on root_dirs and can be modified after encountering a new root_dir but before it goes to `start_client()` via `on_new_config(config)`. In there, you can modify the configuration before it is used to create a client. Many existing clients have examples of that, especially for installed servers via npm.

But amazingly, you've hit the one thing I actually don't allow. I figured that changing initialize options wouldn't be required because rootUri was added a while ago. And workspaceFolders was something I didn't add support for. Seems like I'll be addressing these in https://github.com/neovim/neovim/pull/11430.

As for the first one, you can use the `callbacks` method available on either the skeleton here on `default_config.callbacks` which is to allow setting defaults for a server which are passed to the `config.callbacks` for `start_client(config)`.

**Comment by norcalli** [comment]
I'll see about adding the option to `start_client`. I'd probably do something like `before_initialization = function(initialize_params)` which will let you do various things right before you initialize like modifying the initialize_params.

**Comment by pwntester** [comment]
Thanks, I gave `callbacks` a try but Im probably doing something wrong since Im not able to hijack `publishDiagnostics` message. I tried something along the lines:

```
function! s:Test(error, method, params, client_id) abort
    echom "FOO ".a:method
endfunction

call lsp#add_filetype_config({
    \ 'filetype': 'java',
    \ 'name': 'jdtls',
    \ 'cmd': 'jdtls',
    \ 'callbacks': {
    \   'textDocument/publishDiagnostics': function('<SID>Test') 
    \    }
    \ })
```

But I still get the diagnostics shown and no messaged echoed so it seems my callback was not called.

**Comment by pwntester** [comment]
Oh, I just read this in the docs:

```
 NOTE: the function options in {config} like {config.on_init} are for Lua
  callbacks, not Vim callbacks.
```

So seems I will need to take the lua road after all ...

**Comment by norcalli** [comment]
@pwntester Yes, all roads lead to Lua. :P

**Comment by norcalli** [comment]
@pwntester added `before_init` in https://github.com/neovim/neovim/pull/11430

**Comment by pwntester** [comment]
Awesome, thank you! I think codeActions support is the only missing piece for my setup

**Comment by justinmk** [comment]
@pwntester are you waiting on a change in nvim or nvim-lsp, or can we close this?

> So seems I will need to take the lua road after all ...

The Vimscript wrappers are going to be removed. It's easy to use Lua from Vimscript, see `:help lua-heredoc`.

**Comment by pwntester** [comment]
> The Vimscript wrappers are going to be removed. It's easy to use Lua from Vimscript, see :help lua-heredoc.

Thanks for the hint, I had some trouble finding out how to use lua from vim

Yep, this issue can be closed, thanks! 

Using lua I was able to extend built-in support to do the following:

- add signs to sign column for diagnostics
- expose diagnostic counts with a global function so they can be showed in the status bar
- expose LSP server readyness with a global function so they can be showed in the status bar
- right align virtual text to the right of the screen
- change/remove highlights from diagnostics popup window
- set workspaceFolders initialize param for servers that require it
- handle java ls unsupported notifications such as `language/status`

![imagen](https://user-images.githubusercontent.com/125701/69792750-7c4ff380-11c7-11ea-858a-bb1775e047f3.png)

I dont know if any of these may be of interest for the community, and considering my lua is horrible at the moment, I dont feel comfortable to submit any PRs, but maybe you want to take some of these into consideration.

Still have some items that I want to complete when time permits, espicially codeAction support. It seems that the callbacks are in place to perform de workspace edits, so we just need to implement the request for actions.

In a side note, there are a couple of things that I was not able to do:
- use lua bindings to place signs. I tried `vim.api.sign_define`, etc. but they are not defined so ended up using `vim.api.nvim_command` instead
- implement a catch-all callback for any server notifications. I was able to see that java server was sending these out-of-spec `language/status` notiifcations to the client by using a different client. But I would like to enable a callback to respond to any notification, just for debugging purposes and to see if there is any notification that could be handled and is not.
- refresh diagnosticts on windows resize. a side effect of right-aligned virtual text is that if you open a split (eg: defx in the screenshot above) the virtual text may get out of screen. Is there any autocommand triggered when a window changes its szie? 

My config can be found [here](https://github.com/pwntester/dotfiles/blob/master/config/nvim/lua/lsp/init.lua). As you can see, I had to replicate a lot of neovim code to change the virtual text part, could it be done differently? also any feedback about the code is welcome, Im totally new to lua :)

**Comment by justinmk** [comment]
> Using lua I was able to extend built-in support to do the following:

That's great! We are eager to add sample code to the docs, so please share those in a separate issue.

>  I tried `vim.api.sign_define`, etc. but they are not defined so ended up using `vim.api.nvim_command` instead

`sign_define` is an editor function, so it is exposed to Lua via `vim.fn` . Where `vim.api` exposes the Nvim API. See also `:help vim.fn` .

> implement a catch-all callback for any server notifications

good idea, can you create a separate issue for that. 

Something like `lsp.on_response(handled)` which is `nil` by default, otherwise all server => client responses pass through it, and `handled=true` if there is a handler for it already.

> refresh diagnosticts on windows resize

Yeah, Nvim needs a "WinResized" event...

**Comment by pwntester** [comment]
> sign_define is an editor function, so it is exposed to Lua via vim.fn . Where vim.api exposes the Nvim API. See also :help vim.fn .

thanks! that also helped me with missing `has`, `exists`, etc.. :)

As for opening issues for docs, catch-all handler and WinResized, should I open them in the neovim repo or here?

**Comment by justinmk** [comment]
> As for opening issues for docs, catch-all handler and WinResized, should I open them in the neovim repo or here?

Most of those might belong in the core repo. We can always transfer if not.

**Comment by pwntester** [comment]
will do, thanks

last question about lua interoperability. how can I pass a lua callback to a command run with nvim_command? Im trying something like:
```
vim.api.nvim_command(string.format('call fzf#run(fzf#wrap({"source": %s, "sink": "v:lua.codeactions_callback", "options": "+m --with-nth 2.. -d ::"}))', str_list))
```
but lua function codeactions_callback is never called

also, Im converting a lua table (list) to a string (`str_list`) with:

```
    local str_list = '['
    for idx, option in ipairs(options) do
        str_list = str_list..string.format('"%d::%s",', idx, option.title)
    end
    str_list = str_list..']'
```

Is there any any way of invoking a vimL function with a lua object as argument (in this case a table)?

Thanks!

**Comment by justinmk** [comment]
> how can I pass a lua callback to a command run with nvim_command?

Not supported.

> Is there any any way of invoking a vimL function with a lua object as argument (in this case a table)?

Yes, see `:help vim.fn`. You definitely don't need to do the horrible string concatenation :)

**Comment by clason** [comment]
@pwntester 
> * expose LSP server readiness with a global function so they can be shown in the status bar
> * right align virtual text to the right of the screen

I think both of these would to upstream!
